### DISCUSSION TITLE
Vídeo 04 - Desenvolvimento

### DISCUSSION THREAD

**Discussion by thpoiani** [thpoiani/lab-quarkus#5] [General]
Além de desenvolver, eu também gosto muito de me envolver com Arquitetura e Engenharia de Software 😄 
E finalmente colocamos a mão na massa!! Muitos conceitos foram apresentados nesse vídeo: [Domain Model](https://martinfowler.com/eaaCatalog/domainModel.html), [Java Records](https://docs.oracle.com/en/java/javase/17/language/records.html), [TDD](https://martinfowler.com/bliki/TestDrivenDevelopment.html), [Service Layer](https://martinfowler.com/eaaCatalog/serviceLayer.html), [Dependency Injection](https://www.martinfowler.com/articles/injection.html), [Repository](https://martinfowler.com/eaaCatalog/repository.html), [Query Object](https://martinfowler.com/eaaCatalog/queryObject.html) e [Builder Pattern](https://martinfowler.com/dslCatalog/constructionBuilder.html).

Montamos um serviço simples que atende nossas necessidades iniciais e segue um fundamento essencial: **baixo acoplamento e alta coesão**. E tudo isso foi desenvolvimento com **continuous testing**.

Compartilhe comigo:

- quais desses conceitos são novidades para você?
- você tem o costume de desenvolver testes unitário automatizados?
- quais são as vantagens de tem um processo com teste contínuo?

**Comment by didifive** [comment]
Muito boa a parte dos TDD apresentada, é praticamente uma aula de TDD com baby steps. O legal do TDD é já ir construindo o teste junto com o código, assim quando finaliza a implementação do negócio, já tem o teste pronto. No começo pode parecer complicado para quem já está acostumado iniciar com o código e não tinha treinado testes assim, mas não tem segredo, é praticar.

**Comment by Lucas-Henschel** [comment]
Boa noite Thiago, no final do vídeo onde você gerou o build, quais containers estavam rodando ?

**Reply by thpoiani** [reply]
Olá Lucas, o comando `./mvnw package` pode ser usado para o build da branch `video04`.
Acredito que nesse ponto não é necessário nenhum container sendo executado.
Me avise se tiver algum problema na execução.

**Comment by vicssb** [comment]
Olá!
Compartilhando contigo:

quais desses conceitos são novidades para você?
[Java Records]
você tem o costume de desenvolver testes unitário automatizados?
Sim.
quais são as vantagens de tem um processo com teste contínuo?
Ganho de produtividade, os erros são achados e corrigidos na menor instância do projeto.

**Comment by ale-fuentes-ar** [comment]
## localhost:8080/dashboard refused | Aula 4
<img src="https://img.shields.io/badge/by-Alejandro.Fuentes-informational?style=flat-square&logoColor=white&color=cdcdcd" />

Boa noite team...e boa noite @thpoiani .. novamente grato por compartir e pela sua atenção.

Estou seguindo as aulas, e no passo de navegar na url `http://loclahost:8080/dashboard` não estou tendo resultado

Alguem poderia me indicar alguma dica de como solucinar ?
Gratidão.. e um ótimo dia !!!


![image](https://github.com/thpoiani/lab-quarkus/assets/108128617/368a1b11-705f-4d4b-acd6-2e5102ace21c)

**Reply by thpoiani** [reply]
Olá @ale-fuentes-ar :wave: Normalmente o `localhost` é acessível através do endereço de loopback `127.0.0.1`.
No seu comentário há dois typos: `loclahost` e `127.0.0.0`. Você poderia verificar novamente?

**Reply by ale-fuentes-ar** [reply]
Boa noite @thpoiani .. eu errei mesmo.. era o `127.0.0.1` .. mas segue a mesma resposta ...  :-(

**Comment by ale-fuentes-ar** [comment]
## Erro nos teste | Aula 4
<img src="https://img.shields.io/badge/by-Alejandro.Fuentes-informational?style=flat-square&logoColor=white&color=cdcdcd" />

Segui o tutorial ... o codigo esta igual a branch video 04, mas estou tendo estes impedimentos.

```bash
2024-04-19 20:09:28,619 ERROR [io.qua.test] (Test runner thread) >>>>>>>>>>>>>>>>>>>> Summary: <<<<<<<<<<<<<<<<<<<<
SQLCandidateRepositoryTest#findByName() Cannot invoke "domain.CandidateRepository.save(java.util.List)" because the return value of "domain.CandidateRepositoryTest.repository()" is null
SQLCandidateRepositoryTest#findAll() Cannot invoke "domain.CandidateRepository.save(java.util.List)" because the return value of "domain.CandidateRepositoryTest.repository()" is null
SQLCandidateRepositoryTest#save() Cannot invoke "domain.CandidateRepository.save(domain.Candidate)" because the return value of "domain.CandidateRepositoryTest.repository()" is null
```

Alguem poderia me indicar alguma dica de como solucinar ?
Gratidão.. e um ótimo dia !!!

**Reply by thpoiani** [reply]
Olá @ale-fuentes-ar , acredito que os testes para CandidateRepository acontecem no vídeo 05. De toda forma, confira se o SQLCandidateRepository possui implementação para o método `save`.

Segue a classe para verificação:
- https://github.com/thpoiani/lab-quarkus/blob/video05/election-management/src/main/java/infrastructure/repositories/SQLCandidateRepository.java
- https://github.com/thpoiani/lab-quarkus/blob/video05/election-management/src/test/java/infrastructure/repositories/SQLCandidateRepositoryTest.java
- https://github.com/thpoiani/lab-quarkus/blob/video05/election-management/src/test/java/domain/CandidateRepositoryTest.java

**Comment by mauromortoni** [comment]
Quais são as vantagens de tem um processo com teste contínuo?

No contexto de desenvolvimento de software e práticas ágeis traz diversas vantagens. Por exemplo:

. Detecção Precoce de Erros: testes contínuos permitem identificar e corrigir falhas mais rapidamente, antes que se tornem problemas maiores.

. Feedback Rápido: os desenvolvedores recebem feedback imediato sobre o código que estão escrevendo, o que facilita ajustes e melhorias.

. Aumento da Qualidade do Software: a prática constante de testes ajuda a garantir que o software atenda aos requisitos e mantenha um padrão de qualidade alto.

. Redução de Custos: detectar e corrigir problemas nas fases iniciais do desenvolvimento é geralmente mais barato do que consertá-los após a implementação.

. Facilidade de Integração: com testes contínuos, a integração de novas funcionalidades se torna mais tranquila, pois os testes garantem que mudanças não quebrem a funcionalidade existente.

. Melhoria na Colaboração da Equipe: testes contínuos promovem uma cultura de responsabilidade compartilhada entre os membros da equipe, já que todos estão cientes do estado do código.

. Automação: muitos testes podem ser automatizados, o que economiza tempo e permite que os desenvolvedores se concentrem em tarefas mais complexas.

. Documentação Automatizada: os testes funcionais e de unidade atuam como uma forma de documentação viva do sistema, facilitando a compreensão do comportamento esperado do software.

. Melhoria na Satisfação do Cliente: com qualidade superior e entregas mais rápidas, os clientes tendem a ficar mais satisfeitos com o produto final.

. Escalabilidade: processos de teste contínuo são mais fáceis de escalar conforme a equipe e o projeto crescem, mantendo a qualidade em um ambiente dinâmico.

Dessa forma, o processo de teste contínuo não apenas eleva a qualidade das entregas, mas também acelera sua realização. Isso torna essa prática uma ferramenta essencial para equipes que almejam agilidade e eficiência no desenvolvimento de software.
